src/java/picard/vcf/MendelianViolationDetector.java

```diff
@@ -0,0 +1,338 @@
+                if (momGt.isHomRef() && dadGt.isHomRef() && !kidGt.isHomRef()) {
+                    trio.NUM_DIPLOID_DENOVO++;
+                    type = MendelianViolation.Diploid_Denovo;
+                } else if (momGt.isHomVar() && dadGt.isHomVar() && kidGt.isHet()) {
+                    trio.NUM_HOMVAR_HOMVAR_HET++;
+                    type = MendelianViolation.HomVar_HomVar_Het;
+                } else if (kidGt.isHom() && ((momGt.isHomRef() && dadGt.isHomVar()) || (momGt.isHomVar() && dadGt.isHomRef()))) {
+                    trio.NUM_HOMREF_HOMVAR_HOM++;
+                    type = MendelianViolation.HomRef_HomVar_Hom;
+                } else if (kidGt.isHom() && ((momGt.isHom() && dadGt.isHet()) || (momGt.isHet() && dadGt.isHom()))) {
+                    trio.NUM_HOM_HET_HOM++;
+                    type = MendelianViolation.Hom_Het_Hom;
+                } else {
+                    trio.NUM_OTHER++;
+                    type = MendelianViolation.Other;
+                }
+            }
+
+            // Output a record into the family's violation VCF
+            if (type != null) {
+                // Create a new Context subsetted to the three samples
+                final VariantContextBuilder builder = new VariantContextBuilder(ctx);
+                builder.genotypes(ctx.getGenotypes().subsetToSamples(CollectionUtil.makeSet(trio.MOTHER, trio.FATHER, trio.OFFSPRING)));
+                builder.attribute("MV", type.name());
+
+                // Copy over some useful attributes from the full context
+                if (ctx.hasAttribute(VCFConstants.ALLELE_COUNT_KEY))
+                    builder.attribute("AC_Orig", ctx.getAttribute(VCFConstants.ALLELE_COUNT_KEY));
+                if (ctx.hasAttribute(VCFConstants.ALLELE_FREQUENCY_KEY))
+                    builder.attribute("AF_Orig", ctx.getAttribute(VCFConstants.ALLELE_FREQUENCY_KEY));
+                if (ctx.hasAttribute(VCFConstants.ALLELE_NUMBER_KEY))
+                    builder.attribute("AN_Orig", ctx.getAttribute(VCFConstants.ALLELE_NUMBER_KEY));
+
+                // Write out the variant record
+                familyToViolations.get(trio.FAMILY_ID).add(builder.make());
+            }
+        }
+    }
+
+    @Override
+    public Result result() {
+        return new Result(trios, familyToViolations);
+    }
+
+    /** Tests to ensure that a locus is bi-allelic within the given set of samples. */
+    private boolean isVariant(final Genotype... gts) {
+        for (final Genotype gt : gts) {
+            if (gt.isCalled() && !gt.isHomRef()) return true;
+        }
+
+        return false;
+    }
+
+    /** Tests whether the alleles of the offspring are possible given the alleles of the parents. */
+    private boolean isMendelianViolation(final Genotype p1, final Genotype p2, final Genotype off) {
+        final Allele offAllele1 = off.getAllele(0);
+        final Allele offAllele2 = off.getAllele(1);
+
+        for (final Allele a1 : p1.getAlleles()) {
```

Comment:
I may have written this, but I find it a little confusing to look at now.  Perhaps it could just be:
```java
return !((p1.contains(offAllele1) && p2.contains(offAllele2)) || (p2.contains(offAllele1) && p1.contains(offAllele2)))
```